### PR TITLE
feat: add delete task and list tasks manager api with request type and service type.

### PR DESCRIPTION
**Pull request by IRONICBo**
## Description

Add a POST method to start an asynchronous task to clear the cache of the task id on the peer, and a query api to query the status of the asynchronous task and aggregate it into the existing asynchronous task creation logic.
The manager eventually collects the information needed to start the job and creates a GroupJob into the machine to send a message to the distributed task waiting for the scheduler to consume it.
There is already a Job definition that can reuse the job request to support a new Job, define a new JobRequest, and the response reuses the Job response.

Meanwhile, this PR add a delete task job and list tasks job implementation to the scheduler's async job to receive incoming jobs from the manager and call grpc's interface to complete tasks queries and deletions.

## Related Issue

- https://github.com/dragonflyoss/Dragonfly2/issues/2818

## Motivation and Context

This change introduces api definitions for delete task and list task and request responsetype definitions, as well as interfaces under the services package, and introduced internaljob configuration parameters and job calls to asynchronously list or delete peers and task information for associated tasks in the scheduler.

- Design: https://www.yuque.com/baimo/fv9qk8/ed40b9sg7l9crdt4
- Design: https://github.com/dragonflyoss/Dragonfly2/issues/2818#issuecomment-2226748430

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: manager/job/manager_tasks.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	var tasks []machineryv1tasks.Signature
+	for _, signature := range signatures {
+		tasks = append(tasks, *signature)
+	}
+
+	logger.Infof("create manager tasks group %s in queues %v, tasks: %#v", group.GroupUUID, queues, tasks)
+	if _, err := m.job.Server.SendGroupWithContext(ctx, group, 0); err != nil {
+		logger.Errorf("create  manager tasks group %s failed", group.GroupUUID, err)
```
Comment (chlins) [inline manager/job/manager_tasks.go:121]: Remove extra spaces in log.

File: manager/types/job.go [hunk 2]
```diff
@@ -57,6 +57,34 @@ type CreatePreheatJobRequest struct {
...
+	Type                string         `json:"type" binding:"required"`
+	Args                ListTasksArgs  `json:"args" binding:"omitempty"`
+	Result              map[string]any `json:"result" binding:"omitempty"`
+	UserID              uint           `json:"user_id" binding:"omitempty"`
+	SchedulerClusterIDs []uint         `json:"scheduler_cluster_ids" binding:"omitempty"`
+}
+
+type ListTasksArgs struct {
+	TaskID  string `json:"task_id" binding:"required"`
+	Page    int    `json:"page" binding:"omitempty,gte=1"`
```
Comment (chlins) [inline manager/types/job.go:84]: `Page` and `PerPage` are required by many requests. We can define a common request argument to reuse, but that's not the focus of this PR. We can address it in a subsequent refactoring.

Comment (IRONICBo) [reply manager/types/job.go:84]: Well, there is a `page` and `perpage` parameter in the task itself, but this is mainly for handling `page` and `perpage` in the structure returned within the task result, and I'll consider defining a generic version in subsequent implementations as well.

File: manager/job/manager_tasks.go (hunk 1 above)
Comment (IRONICBo) [reply manager/job/manager_tasks.go:121]: Ok, done.

File: scheduler/job/job.go [hunk 3]
```diff
@@ -297,3 +303,118 @@ func (j *job) syncPeers() (string, error) {
 
 	return internaljob.MarshalResponse(hosts)
 }
+
+// listTasks is a job to list tasks.
+func (j *job) listTasks(ctx context.Context, data string) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, listTasksTimeout)
```
Comment (chlins) [inline scheduler/job/job.go:309]: I didn't see any reference or usage for the `ctx` in this method.

File: scheduler/job/job.go [hunk 4]
```diff
@@ -297,3 +303,118 @@ func (j *job) syncPeers() (string, error) {
...
+		Page:  req.Page,
+		Peers: peers[req.Page*req.PerPage : (req.Page+1)*req.PerPage],
+	}
+
+	return internaljob.MarshalResponse(listTaskResponse)
+}
+
+// deleteTask is a job to delete task.
+func (j *job) deleteTask(ctx context.Context, data string) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, deleteTaskTimeout)
```
Comment (chlins) [inline scheduler/job/job.go:342]: ditto

File: scheduler/job/job.go (hunk 3 above)
Comment (IRONICBo) [reply scheduler/job/job.go]: Ok, this file should belong to another scheduler PR, which I have currently removed.

File: scheduler/job/job.go (hunk 4 above)
Comment (IRONICBo) [reply scheduler/job/job.go]: Ok.

**Review by chlins — APPROVED** [review]
lgtm

File: scheduler/job/job.go [hunk 5]
```diff
@@ -297,3 +314,163 @@ func (j *job) syncPeers() (string, error) {
...
+	deleteTaskResponse := &internaljob.DeleteTaskResponse{
+		SuccessTasks: successTasks,
+		FailureTasks: failureTasks,
+	}
+
+	return internaljob.MarshalResponse(deleteTaskResponse)
+}
+
+// getValidPeers try to get valid peers by task id
+func (j *job) getValidPeers(taskID string) ([]*resource.Peer, error) {
```
Comment (gaius-qi) [inline scheduler/job/job.go:447]: getFinishedPeers

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/job/job.go [hunk 6]
```diff
@@ -297,3 +314,163 @@ func (j *job) syncPeers() (string, error) {
...
+	// get task info by task id
+	task, ok := j.resource.TaskManager().Load(taskID)
+	if !ok {
+		logger.Errorf("task %s not found", taskID)
+		return nil, fmt.Errorf("task %s not found", taskID)
+	}
+
+	// get peer info by task info
+	peers := make([]*resource.Peer, 0)
```
Comment (gaius-qi) [inline scheduler/job/job.go:456]: Add `LoadFinishedPeers` to task.

File: scheduler/job/job.go [hunk 7]
```diff
@@ -297,3 +314,163 @@ func (j *job) syncPeers() (string, error) {
...
+		return "", err
+	}
+
+	// Delete task by task id and host id
+	successTasks := make([]*internaljob.Task, 0)
+	failureTasks := make([]*internaljob.Task, 0)
+
+	// Create a wait group to limit delete rpc concurrency
+	// and avoid too many rpc requests to the host.
+	wg := sync.WaitGroup{}
```
Comment (gaius-qi) [inline scheduler/job/job.go:374]: No concurrent deletion.

File: scheduler/job/job.go [hunk 8]
```diff
@@ -297,3 +314,163 @@ func (j *job) syncPeers() (string, error) {
...
+		deleteTaskLimit <- struct{}{}
+		go func(peer *resource.Peer) {
+			defer func() {
+				wg.Done()
+				<-deleteTaskLimit
+			}()
+
+			// Get dfdaemon client from host
+			target := fmt.Sprintf("%s:%d", peer.Host.IP, peer.Host.Port)
+			conn, err := grpc.DialContext(
```
Comment (gaius-qi) [inline scheduler/job/job.go:387]: Use `pkg/rpc` client.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/job/job.go (hunk 5 above)
Comment (IRONICBo) [reply scheduler/job/job.go:447]: OK, I have updated this function's name.

File: scheduler/job/job.go (hunk 6 above)
Comment (IRONICBo) [reply scheduler/job/job.go:456]: OK, moved to `scheduler/resource/task.go::LoadFinishedPeers`.

File: scheduler/job/job.go (hunk 7 above)
Comment (IRONICBo) [reply scheduler/job/job.go:374]: OK, removed.

File: scheduler/job/job.go (hunk 8 above)
Comment (IRONICBo) [reply scheduler/job/job.go:387]: OK, I have added a `GetV2ByAddr` in `pkg/rpc/dfdaemon/client/client_v2.go`, refer to `pkg/rpc/manager/client/client_v2.go::GetV2ByAddr()`

**Review by gaius-qi — APPROVED** [review]
LGTM

**Comment by gaius-qi** [comment]
@IRONICBo Fix Lint.

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**